packages/tpetra/core/src/Tpetra_Import_Util2.hpp

```diff
@@ -946,6 +952,287 @@ lowCommunicationMakeColMapAndReindex (const Teuchos::ArrayView<const size_t> &ro
+  // most INT_MAX entries, but it's possible to have more rows than
+  // that (if size_t is 64 bits and int is 32 bits).
+  Kokkos::UnorderedMap<LO, bool, execution_space> LocalGIDs_view_map(colind_LID.size());
+  Kokkos::UnorderedMap<GO, LO, execution_space> RemoteGIDs_view_map(colind_LID.size());
+  
+  const size_t numMyRows = rowptr.size () - 1;
+  local_map_type domainMap_local = domainMap.getLocalMap();
+  
+  const size_t numDomainElements = domainMap.getLocalNumElements ();
+  Kokkos::View<bool*> LocalGIDs_view("LocalGIDs", numDomainElements);
+  auto LocalGIDs_host = Kokkos::create_mirror_view(LocalGIDs_view);
+  
+  size_t NumLocalColGIDs = 0;
+  
+  // Scan all column indices and sort into two groups:
+  // Local:  those whose GID matches a GID of the domain map on this processor and
+  // Remote: All others.
+  // Kokkos::Parallel_reduce sums up NumLocalColGIDs, while we use the size of the Remote GIDs map to find NumRemoteColGIDs
+  Kokkos::parallel_reduce(team_policy(numMyRows, Kokkos::AUTO), KOKKOS_LAMBDA(const typename team_policy::member_type &member, size_t &update) { 
+    const int i = member.league_rank();
+    size_t NumLocalColGIDs_temp = 0;
+    size_t rowptr_start = rowptr_view[i];
+    size_t rowptr_end = rowptr_view[i+1];
+    Kokkos::parallel_reduce(Kokkos::TeamThreadRange(member, rowptr_start, rowptr_end), [&](const size_t j, size_t &innerUpdate) {
+      const GO GID = colind_GID_view[j];
+      // Check if GID matches a row GID in local domain map
+      const LO LID = domainMap_local.getLocalElement (GID);
+      if (LID != -1) {
+        auto outcome = LocalGIDs_view_map.insert(LID);
+        // Fresh insert
+        if(outcome.success()) {
+          LocalGIDs_view[LID] = true;
+          Kokkos::atomic_add(&innerUpdate, 1);
+        }
+      }
+      else {
+        const int PID = owningPIDs_view[j];
+        auto outcome = RemoteGIDs_view_map.insert(GID, PID);
+        if(outcome.success() && PID == -1) {
+          printf("Cannot figure out if ID is owned.\n");
+          Kokkos::abort("Cannot figure out if ID is owned.\n");
+        }
+      }
+    }, NumLocalColGIDs_temp);
+    if(member.team_rank() == 0) update += NumLocalColGIDs_temp;
+  }, NumLocalColGIDs);
+  
+  LO NumRemoteColGIDs = RemoteGIDs_view_map.size();
+
+  // For each index in RemoteGIDs_map that contains a GID, use valid_key_index[i] to indicate the number of GIDs "before" this GID
+  // Each valid_key_index[i], where i is an index in RemoteGIDs_map, is unique and consecutive
+  // This maps each element in the RemoteGIDs hash table to an index in RemoteGIDList / PIDList without any overwrites or empty spaces
+  Kokkos::View<int*> valid_key_index("valid_key_index", RemoteGIDs_view_map.capacity()); 
+  Kokkos::parallel_scan(RemoteGIDs_view_map.capacity(), KOKKOS_LAMBDA(const int i, GO& update, const bool final) {
+    if(final && RemoteGIDs_view_map.valid_at(i)) {
+      valid_key_index[i] = update;
```

Comment:
You can actually merge this with the parallel_for below (and not use ``valid_key_index``). ``final`` will be true exactly once for each index, so it's safe to do these here:
```
RemoteGIDList_view(update) = RemoteGIDs_view_map.key_at(i);
PIDList_view(update) = RemoteGIDs_view_map.value_at(i);
```